src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs

```diff
@@ -490,8 +515,33 @@ public bool TryGetDateTimeOffset(out DateTimeOffset value)
                 throw ThrowHelper.GetInvalidOperationException_ExpectedString(TokenType);
             }
 
-            ReadOnlySpan<byte> span = HasValueSequence ? ValueSequence.ToArray() : ValueSpan;
-            return JsonHelpers.TryParseAsISO(span, out value, out int bytesConsumed) && span.Length == bytesConsumed;
+            ReadOnlySpan<byte> span = ValueSpan;
+
+            if (HasValueSequence)
+            {
+                JsonReaderHelper.GetValueSequenceAsSpan(ValueSequence, out span);
+            }
+
+            int bytesConsumed;
+
+            if (_stringHasEscaping)
+            {
+                int idx = span.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(idx != -1);
+
+                Span<byte> utf8Unescaped = stackalloc byte[span.Length];
+
+                JsonReaderHelper.Unescape(span, utf8Unescaped, idx, out int written);
+                Debug.Assert(written > 0);
+
+                utf8Unescaped = utf8Unescaped.Slice(0, written);
+                Debug.Assert(!utf8Unescaped.IsEmpty);
+
+                return JsonHelpers.TryParseAsISO(utf8Unescaped, out value, out bytesConsumed) && utf8Unescaped.Length == bytesConsumed;
+            }
+
+            Debug.Assert(span.IndexOf(JsonConstants.BackSlash) == -1);
+            return JsonHelpers.TryParseAsISO(span, out value, out bytesConsumed) && span.Length == bytesConsumed;
```

Comment:
Can we do the same length checks here, like `TryGetGuid`?

Reply:
We'd need to decide on the maximum number of fraction digits allowed. Any thoughts?

Reply:
I've capped the allowed number of fraction digits to 9, given use cases like https://github.com/dotnet/corefx/issues/36023#issuecomment-472608096.
@JamesNK @bartonjs @rynowak - can you confirm if you are fine with this?

Reply:
If we do 7 and Java does 9, why not parse out to like 15?  It's still small enough that it's within reason for stackalloc escaping, but large enough that we won't get a "oh, but [some platform] does 10,". (This is just the usual "figure out the biggest "reasonable" value, then add { 10, 25, 50, 100, 100000 }% (depending on what you can accept) of "I don't want to argue" margin)

Random information:
The absolute biggest (under current interpretation of physics) that anyone could legitimately claim is 44 digits ([Planck time](https://en.wikipedia.org/wiki/Planck_time)).  But that seems a bit far. -- The article claims an experiment has measured something in the 19-digit range; following the reference suggests they had 21-digit precision on their measurements.

Reply:
I think 15 is reasonable for the reason you mentioned. I'll update to this if there are no other thoughts.